File: packages/cli/src/verify/scopes/decision-projection-spike.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { assert, fail } from "../check.js";
import type { CheckResult, Scope, ScopeContext } from "../types.js";

const RESULTS_REL = "core/build/decision-projection-results.json";
const JVM_COMMAND = "cd core && ./gradlew runDecisionSpike";

interface VectorResult {
  id: string;
  expect: string | null;
  reference: string | null;
  compiled: string | null;
  match: boolean;
  referenceErrors: boolean;
  compiledErrors: boolean;
}

interface Results {
  engine: string;
  specVersion: string;
  decision: string;
  vectors: VectorResult[];
}

/** Numeric-tolerant equality so FEEL BigDecimal scale (`50` vs `50.0`) is not a spurious mismatch. */
function valueMatches(a: string | null, b: string | null): boolean {
  if (a === null || b === null) return false;
  if (a === b) return true;
  const na = Number(a);
  const nb = Number(b);
  return !Number.isNaN(na) && !Number.isNaN(nb) && na === nb;
}

/**
 * decision-projection-spike — build plan 2.0 (ADR-0001/0002/0027), the DMN-projection feasibility
 * gate. Consumes the KIE/Drools execution results (`cd core && ./gradlew runDecisionSpike`) where the
 * `decision-source` fixture is compiled to DMN 1.6 and run alongside the hand-authored reference. Per
 * vector it asserts: neither model errored, the compiled projection matches the reference, and both
 * match the expected outcome. This is the proof that the hard boxed-expression kinds (BKM FEEL
 * functions, boxed contexts, invocations) project and execute correctly before the full compiler.
 */
export const decisionProjectionSpikeScope: Scope = {
  id: "decision-projection-spike",
  description:
    "Compiled decision-source → DMN 1.6 executes identically to the hand-authored reference on KIE/Drools across every input vector (hard boxed-expression kinds).",
  run({ repoRoot }: ScopeContext): CheckResult[] {
    const resultsPath = join(repoRoot, RESULTS_REL);
    if (!existsSync(resultsPath)) {
      return [
        fail("decision-projection-spike.results-present", {
          diff: `missing ${RESULTS_REL}; run \`${JVM_COMMAND}\` to compile and execute on KIE`,
        }),
      ];
    }

    const results = JSON.parse(readFileSync(resultsPath, "utf8")) as Results;
    const checks: CheckResult[] = [];

    checks.push(
      assert("decision-projection-spike.engine", results.engine.startsWith("kie-dmn:"), {
        expected: "kie-dmn:<version>",
        actual: results.engine,
      }),
    );

    for (const v of results.vectors) {
      checks.push(
        assert(
          `decision-projection-spike.no-errors.${v.id}`,
          !v.referenceErrors && !v.compiledErrors,
          {
            expected: "no DMN errors on reference or compiled",
            actual: `reference=${v.referenceErrors}, compiled=${v.compiledErrors}`,
          },
        ),
      );
      checks.push(
        assert(`decision-projection-spike.match.${v.id}`, v.match, {
          expected: `compiled == reference (${v.reference})`,
          actual: `compiled=${v.compiled}`,
        }),
      );
      checks.push(
        assert(`decision-projection-spike.expected.${v.id}`, valueMatches(v.reference, v.expect), {
          expected: v.expect ?? "(none)",
          actual: v.reference,
        }),
      );
    }

    return checks;
  },
};

File: core/src/main/kotlin/ai/ichiflow/core/decision/DecisionProjectionSpike.kt
package ai.ichiflow.core.decision

import ai.ichiflow.core.decision.spi.DecisionEngine
import ai.ichiflow.core.decision.spi.DroolsDecisionEngine
import ai.ichiflow.core.decision.spi.LoadedModel
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import java.io.File

/**
 * Phase 2.0 decision-projection spike (build plan 2.0, ADR-0001/0002/0027). Compiles the
 * `decision-source` fixture to DMN 1.6 and executes **both** it and the hand-authored reference DMN
 * on Apache KIE / Drools (pinned 10.2.0) across the input vectors, asserting identical results. This
 * retires "riskiest bet #2": the hard boxed-expression kinds (BKM FEEL functions, boxed contexts,
 * invocations) both compile to valid DMN *and* execute identically to a reference. Results are
 * written machine-readably for the `decision-projection-spike` verify scope.
 *
 * As of Phase 2.1 the KIE coupling lives behind the [DecisionEngine] SPI; this spike drives that
 * engine rather than talking to `org.kie.*` directly.
 */
object DecisionProjectionSpike {

    /** Evaluate one decision over one input row; returns the stringified result + whether it errored. */
    fun evaluate(engine: DecisionEngine, model: LoadedModel, inputs: Map<String, Any?>, decisionName: String): Pair<String?, Boolean> {
        val evaluation = engine.evaluate(model, inputs)
        return evaluation.resultOf(decisionName)?.toString() to evaluation.hasErrors
    }

    @JvmStatic
    fun main(args: Array<String>) {
        // Working dir is core/; the decision-source + vectors live one level up under schemas/.
        val repoRoot = File("..").canonicalFile
        val sourceFile = File(repoRoot, args.getOrElse(0) { "schemas/decision-source/fixtures/fee.decision.json" })
        val vectorsFile = File(repoRoot, args.getOrElse(1) { "schemas/decision-source/vectors.json" })
        val outputFile = File(args.getOrElse(2) { "build/decision-projection-results.json" })

        val mapper = ObjectMapper().registerKotlinModule()
        val source = mapper.readValue(sourceFile, DecisionSource::class.java)
        val vectors = mapper.readTree(vectorsFile)
        val decisionName = vectors["decision"].asText()

        val referenceXml = javaClass.getResourceAsStream("/dmn/reference/fee.dmn")!!
            .bufferedReader().use { it.readText() }
        val compiledXml = DecisionSourceCompiler.compile(source)

        val engine = DroolsDecisionEngine()
        val referenceModel = engine.load(referenceXml, "reference.dmn")
        val compiledModel = engine.load(compiledXml, "compiled.dmn")

        val root = mapper.createObjectNode()
        root.put("engine", "kie-dmn:${engine.capabilities.engineVersion}")
        root.put("specVersion", "DMN 1.6")
        root.put("decision", decisionName)
        val arr = root.putArray("vectors")

        for (vector in vectors["vectors"]) {
            val id = vector["id"].asText()
            val inputs = mutableMapOf<String, Any?>()
            vector["inputs"].fields().forEach { (k, node) ->
                inputs[k] = if (node.isNumber) node.numberValue() else node.asText()
            }
            val (refValue, refError) = evaluate(engine, referenceModel, inputs, decisionName)
            val (compValue, compError) = evaluate(engine, compiledModel, inputs, decisionName)

            val node = arr.addObject()
            node.put("id", id)
            node.put("expect", vector["expect"]?.asText())
            node.put("reference", refValue)
            node.put("compiled", compValue)
            node.put("match", refValue != null && refValue == compValue)
            node.put("referenceErrors", refError)
            node.put("compiledErrors", compError)
        }

        outputFile.parentFile.mkdirs()
        outputFile.writeText(mapper.writerWithDefaultPrettyPrinter().writeValueAsString(root) + "\n")
        println("Wrote ${vectors["vectors"].size()} projection results to ${outputFile.path}")
    }
}
